lib/IDE/Refactoring.cpp

```diff
@@ -2066,6 +2067,160 @@ bool RefactoringActionSimplifyNumberLiteral::performChange() {
   return true;
 }
 
+static CallExpr *findTrailingClosureTarget(SourceManager &SM,
+                                           ResolvedCursorInfo CursorInfo) {
+  if (!CursorInfo.ContextNode)
+    // Middle of nowhere.
+    return nullptr;
+
+  SourceLoc Loc;
+  switch (CursorInfo.Kind) {
+  case CursorInfoKind::ExprStart:
+    Loc = CursorInfo.TrailingExpr->getStartLoc();
+    break;
+  case CursorInfoKind::StmtStart:
+    return nullptr;
+  default:
+    Loc = CursorInfo.Loc;
+    break;
+  }
+
+  assert(Loc.isValid() && "Invalid loc to find");
+
+  // Find inner most CallExpr
+  struct CallExprFinder : public SourceEntityWalker {
+    SourceManager &SM;
+    SourceLoc Loc;
+    CallExpr *Target = nullptr;
+
+    bool walkToDeclPre(Decl *D, CharSourceRange Range) override {
+      return true;
+    }
+
+    bool walkToStmtPre(Stmt *S) override {
+      if (SM.rangeContainsTokenLoc(S->getSourceRange(), Loc))
+        return false;
+      if (isa<BraceStmt>(S))
+        // This happens if the cursor is on spaces in a closure.
+        Target = nullptr;
+      return true;
+    }
+
+    bool walkToExprPre(Expr *E) override {
+      if (!E->getSourceRange().isValid())
+        return false;
+      if (!SM.rangeContainsTokenLoc(E->getSourceRange(), Loc))
+        return false;
+      if (isa<ClosureExpr>(E))
+        // If the cursor is on '{' or '}', don't dig into it.
+        return Loc != E->getEndLoc() && Loc != E->getStartLoc();
+      if (auto *CE = dyn_cast<CallExpr>(E))
+        Target = CE;
+      return true;
+    }
+
+    bool walkToExprPost(Expr *E) override {
+      return false;
+    }
+
+  public:
+    CallExprFinder(SourceManager &SM, SourceLoc Loc) : SM(SM), Loc(Loc) {}
```

Comment:
Once we walk the entire source file to find the call expression. We can use an existing utility in this file `ContextFinder`. right?